### DISCUSSION TITLE
How to monitor external resources periodically

### DISCUSSION THREAD

**Discussion by anekdoti** [dotnet/dotnet-operator-sdk#277] [Q&A] [answered]
To monitor external resources (like databases) in the controller, e.g., to update the status of a controlled entity accordingly, events need to be requeued periodically. For example, by requeueing events when leaving the handler methods in the controller with 10 seconds delay, we can be sure that a handler is called at least every 10 seconds. 

However, it seems that this way one is collecting more and more events in the queue of requeued events, which each are requeued again every 10 seconds. For example, if there were updates to the controlled entity at 10:00:01, 10:00:02, ... 10:00:09, then we will have approriate handler methods in the controller called every second forever. 

What is the best way to requeue events to allow periodical actions without ending in a scenario as described above? Or do I misunderstand the event handling?

**Comment by erin-allison** [comment, marked as answer]
There's a couple of different (fairly easy) options that I see here:

1. Run a separate IHostedService with a timer that checks your external resources on whatever interval you desire.
2. (Ab)use the fact that the SDK's resource watcher reconnects every 60 seconds and will process a notmodified event for all watched resources at that time, and use that event to check status of the external resources.

**Comment by buehler** [comment]
In addition to what @erin-allison proposed:

What you describe is the idea of https://github.com/smartive/hosted-database-operator/. In https://github.com/smartive/hosted-database-operator/blob/master/src/HostedDatabaseOperator/Controller/HostController.cs#L42-L58 we check the connection on "created", "updated", and "not-modified" events and requeue them with 1 minute delay.

Then the status / events are updated on the entity.

_But_: Now if I'm thinking about, you are correct. The requeue mechanism does not check if the given resource (uuid) is already requeued which could lead to such a situation where you pile up requeued entities.

**Reply by anekdoti** [reply]
Thank you very much for the reply, @erin-allison and @buehler. We will consider your proposals, @erin-allison. In regards to your "but", @buehler: currently we use the approach you describe. We played a bit with it to see if we can pile up events - but interestingly, this did not happen. In fact, we got a sequence of events as follows:

Update event at t=0 (triggered by update to the entity)
NotModified at t=10
NotModified at t=20
Update event at t=25 (triggered by update to the entity)
NotModified at t=35
NotModified at t=45
...

Initially, I would have expected to see NotModified events at t=30,t=35,t=40,t=45,t=50, ...

I.e., it appears as if there is some kind of deduplication going on - although I could not found any in the KubeOps sources as well!

**Reply by buehler** [reply]
🙈 well... I need to review my own code then :-) thanks for the experiment!

**Reply by anekdoti** [reply]
In #281 you mentioned that this piling up can only happen when the delay is longer than the watcher reconnect interval (which was not the case in our experiment). Could you elaborate on the difference? Although I try to understand the KubeOps sources, some things are still hard to grasp for me - in particular, I do not have a lot of experience with the reactive extensions.

**Reply by buehler** [reply]
The thing is, that if the requeue delay is *shorter* than the watcher timeout, the queue only handles this particular event.

```
Lets say: watcher timeout: 60s
created event -> requeue in 10s. queue count: 1
after 10s -> event triggers. -> requeue in 10s. queue count: 1
after 20s -> event triggers. -> requeue in 10s. queue count: 1
after 30s -> event triggers. -> requeue in 10s. queue count: 1
after 40s -> event triggers. -> requeue in 10s. queue count: 1
after 50s -> event triggers. -> requeue in 10s. queue count: 1
after 60s: special case.
1. event triggers (with a requeue in) -> queue count 1 (from the old requeue in request)
2. watcher reconnects, which triggers a new event in the controller, that requests a requeue in
--> queue count: 2 (with the same entity)
```

As far as I can tell for now.

The same problem arises when:

```
watcher timeout: 30s
created event triggers -> requeue requested in 60s. queue count: 1
after 30s, the watcher terminates and reconnects, which triggers a non-modified event -> if requeue in requested with 60s -> queue count: 2
after 60s: first event triggers, watcher reconnects -> queue count 3
```

I just reviewed the logic, there is no logic to deduplicate same UUIDs for entities (for now)

An older version used .net events but they are prone to memory leaks. This is the main reason I switched to reactive extensions.

The code that needs refactoring is: https://github.com/buehler/dotnet-operator-sdk/blob/master/src/KubeOps/Operator/Controller/ManagedResourceController%7BTEntity%7D.cs#L65-L116

**Reply by anekdoti** [reply]
Do I understand right, that the piling up therefore just happens because reconnecting the watcher always adds a new event to the queue?

**Reply by buehler** [reply]
Exactly

**Reply by anekdoti** [reply]
Would it be possible to only add a new event during reconnect when there is no event pending? And why is there a new event added at all?

**Reply by buehler** [reply]
I guess somehow this should be possible. The reason for the new event is the watcher. The watcher (from Kubernetes, not the SDK) is initialized and returns a list of "new" objects. So when a watcher is started, all found objects are "created" for the watcher. And when Kubernetes terminates the connection to the watcher and a new one is created, the found objects are "new" to the new watcher.

**Reply by anekdoti** [reply]
Thank you very much for the clarification! But when I understand right, we should be safe when we choose a requeue delay that is shorter than the watcher timeout, right?
